File: java/LRUCache.java
import java.util.Map;
import java.util.HashMap;

public class LRUCache {
    private class CacheNode {
        private int key;
        private int value;

        private CacheNode prev;
        private CacheNode next;

        public CacheNode(int key, int value) {
            this.key = key;
            this.value = value;
            this.prev = null;
            this.next = null;
        }
    }

    private int capacity;
    private int numNodes;

    private CacheNode head;
    private CacheNode tail;

    private Map<Integer, CacheNode> map;

    public LRUCache(int capacity) {
        this.capacity = capacity;
        numNodes = 0;

        head = new CacheNode(0, 0);
        tail = new CacheNode(0, 0);
        head.prev = null;
        head.next = tail;
        tail.prev = head;
        tail.next = null;

        map = new HashMap<Integer, CacheNode>();
    }

    public int get(int key) {
        if (capacity == 0) {
            return -1;
        }
        CacheNode curr = map.get(key);
        if (curr == null) {
            return -1;
        }
        //remove from list
        curr.prev.next = curr.next;
        curr.next.prev = curr.prev;

        //add after head
        curr.next = head.next;
        head.next.prev = curr;
        curr.prev = head;
        head.next = curr;
        return curr.value;
    }

    public void set(int key, int value) {
        if (capacity == 0) {
            return;
        }
        CacheNode curr = map.get(key);
        if (curr != null) {
            curr.prev.next = curr.next;
            curr.next.prev = curr.prev;
            curr.value = value;
            --numNodes;
        } else {
            curr = new CacheNode(key, value);
            map.put(key, curr);
        }
        if (numNodes == capacity) {
            CacheNode temp = tail.prev;
            map.remove(temp.key);
            temp.prev.next = temp.next;
            temp.next.prev = temp.prev;
            --numNodes;
        }
        curr.next = head.next;
        head.next.prev = curr;
        curr.prev = head;
        head.next = curr;
        ++numNodes;
        return;
    }

    public static void main(String[] args) {
        LRUCache solver = null;
        solver = new LRUCache(2);
        solver.set(2, 1);
        solver.set(2, 2);
        System.out.println(solver.get(2));
        solver.set(1, 1);
        solver.set(4, 1);
        System.out.println(solver.get(2));

        solver = new LRUCache(2);
        System.out.println(solver.get(2));
        solver.set(2, 6);
        System.out.println(solver.get(1));
        solver.set(1, 5);
        solver.set(1, 2);
        System.out.println(solver.get(1));
        System.out.println(solver.get(2));
    }
}

File: python/lru_cache.py
#!/usr/bin/env python3

class CacheNode:
    def __init__(self, key, val):
        self.key    = key
        self.val    = val
        self.prev   = None
        self.post   = None

class CacheList:
    def __init__(self):
        self.head       = CacheNode(0, 0)
        self.tail       = CacheNode(0, 0)
        self.head.post  = self.tail
        self.tail.prev  = self.head
        self.num_nodes  = 0

    def add_first(self, node):
        temp = self.head.post
        node.prev = self.head
        node.post = temp
        self.head.post = node
        temp.prev = node
        self.num_nodes += 1

    def delete(self, node):
        node.prev.post = node.post
        node.post.prev = node.prev
        self.num_nodes -= 1

    def delete_last(self):
        node = self.tail.prev
        self.delete(node)
        return node

    def size(self):
        return self.num_nodes

    def empty(self):
        return self.num_nodes == 0

class LRUCache:
    def __init__(self, capacity):
        self.capacity   = capacity
        self.cache_list = CacheList()
        self.track      = {}

    def get(self, key):
        if self.cache_list.empty() or key not in self.track:
            return -1
        node = self.track[key]
        self.cache_list.delete(node)
        self.cache_list.add_first(node)
        return node.val

    def set(self, key, val):
        if self.capacity < 1:
            return
        node = None
        if key not in self.track:
            node = CacheNode(key, val)
            self.track[key] = node
        else:
            node = self.track[key]
            node.val = val
            self.cache_list.delete(node)
        if self.cache_list.size() == self.capacity:
            removed = self.cache_list.delete_last()
            del self.track[removed.key]
        self.cache_list.add_first(node)

def main():
    lru_cache = LRUCache(3)
    lru_cache.set(1, 1)
    lru_cache.set(2, 2)
    lru_cache.set(3, 3)
    lru_cache.set(4, 4)
    print(lru_cache.get(4))
    print(lru_cache.get(3))
    print(lru_cache.get(2))
    print(lru_cache.get(1))
    lru_cache.set(5, 5)
    print(lru_cache.get(1))
    print(lru_cache.get(2))
    print(lru_cache.get(3))
    print(lru_cache.get(4))
    print(lru_cache.get(5))

if __name__ == '__main__':
    main()
